enhancements/network/egress-router.md

```diff
@@ -0,0 +1,204 @@
+
+## Design Details
+
+We'll be implenting this feature using an additional specific CNI plugin for
+it, whose [github repository](https://github.com/openshift/egress-router-cni/)
+is already created.
+
+Both kubernetes' pod and multus' net-attach-def are namespaced objects. Due to a
+security limitation in OCP, Pod yaml and net-attach-def object should be in the
+same namespace, thus egress-router-cni would net to take this into
+consideration and put them in the same namespace as well.
+
+The egress router pod would have multus annotations saying to use the
+egress-router-cni plugin, and then the CNI plugin would create a macvlan
+interface (or, in the future, maybe ipvlan or Amazon ENI).
+
+### Overview
+
+The `egress-router` plugin creates some sort of
+cluster-network-external network interface and assigns a user-provided
+static public IP address to it. It is designed for use by the
+OpenShift egress-router feature, but may have other uses as well.
+
+### Example configuration
+
+#### Network Attachment Definition
+
+```
+apiVersion: "k8s.cni.cncf.io/v1"
+kind: NetworkAttachmentDefinition
+metadata:
+  name: egress-router
+  namespace: egress-router-ns
+spec:
+  config: '{
+    "type": "egress-router",
+    "ip": {
+      "addresses": ["10.10.10.230"],
+      }
+    }'
+```
+
+#### ConfigMap
+
+```
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: egress-file
+  namespace: default
+data:
+  - ip:
+      addresses:
+        -
+      gateway:
+      destinations:
+        -
+```
+
+#### Pod with annotation
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: egress
+  labels:
+    name: egress
+  annotations:
+    k8s.v1.cni.cncf.io/networks: egresss-router-ns/egress-router
+spec:
+  containers:
+  - name: openshift-egress
+    image:
+    securityContext:
+      privileged: true
+  volumes:
+  - name: foo
+    configMap:
+      name: egressconfigmap
+```
+
+### Network configuration reference
+
+The whole set of options can be found at https://github.com/openshift/egress-router-cni/blob/master/README.md
+
+### Interface Types and Platform Support
+
+On bare-metal nodes, `macvlan` and `ipvlan` are supported for `interfaceType`, with `macvlan` being the default. For `macvlan`, `interfaceArgs` can include `mode` and `master`, and for `ipvlan` it can include `master`. However, you do not need to specify `master` if it can be inferred from the IP address. (That is, if there is exactly 1 network interface on the node whose configured IP is in the same CIDR range as the pod's configured IP, then that interface will automatically be used as the `master`, and the associated gateway will automatically be used as the `gateway`.)
+
+On AWS, the default `interfaceType` is `aws-elastic-ip`. The administrator is responsible for allocating the Elastic IP address (and for deallocating it when it is no longer needed). The plugin will handle associating the Elastic IP with the node hosting the pod (and moving it to another node if the pod moves).
```

Comment:
Do we really want to implement this now?

IMHO, we are looking to have feature parity with previous egress-router which had macvlan/ipvlan , so AWS could be implemented later.